### PR TITLE
[TIMOB-20286] iOS/Android: Expose Ti.UI.View.getViewById method

### PR DESCRIPTION
**Pull request by hansemannn**
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-20286

### CODE REVIEW

File: android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java [hunk 1]
```diff
@@ -737,7 +737,32 @@ public void removeAllChildren()
 			}
 		}
 	}
+	
+	/**
+	* Returns the view by the given ID.
+	* @module.api
+	*/
+	@Kroll.method
+	public TiViewProxy getViewById(Object arg)
```
Comment (garymathews) [inline android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java]:
```Java
public TiViewProxy getViewById(String id)
```

File: android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java [hunk 2]
```diff
@@ -737,7 +737,32 @@ public void removeAllChildren()
...
+		if (children != null) {
+			for (TiViewProxy child : children) {
+				if (child.children != null && child.children.size() > 0) {
+					TiViewProxy parentChild = child.getViewById(TiConvert.toString(arg));
+					if (parentChild != null) {
+						return parentChild;
+					}
+				}
+
+				if (child.getProperty(TiC.PROPERTY_ID) != null && child.getProperty(TiC.PROPERTY_ID).equals(TiConvert.toString(arg))) {
```
Comment (garymathews) [inline android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java]:
```Java
if (child.hasProperty(TiC.PROPERTY_ID) && child.getProperty(TiC.PROPERTY_ID).equals(id)) {
```

**Comment by hansemannn** [comment]
@garymathews Updated PR!

File: android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java [hunk 3]
```diff
@@ -737,7 +737,32 @@ public void removeAllChildren()
...
+		if (children != null) {
+			for (TiViewProxy child : children) {
+				if (child.children != null && child.children.size() > 0) {
+					TiViewProxy parentChild = child.getViewById(arg);
+					if (parentChild != null) {
+						return parentChild;
+					}
+				}
+
+				if (child.getProperty(TiC.PROPERTY_ID) && child.getProperty(TiC.PROPERTY_ID).equals(arg)) {
```
Comment (garymathews) [inline android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java]:
```Java
if (child.hasProperty(TiC.PROPERTY_ID) && child.getProperty(TiC.PROPERTY_ID).equals(id)) {
```

File: android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java [hunk 4]
```diff
@@ -737,7 +737,32 @@ public void removeAllChildren()
 			}
 		}
 	}
+	
+	/**
+	* Returns the view by the given ID.
+	* @module.api
+	*/
+	@Kroll.method
+	public TiViewProxy getViewById(String arg)
```
Comment (garymathews) [inline android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java]:
```Java
String id // as documented
```

Comment (hansemannn) [inline android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java]: Does this work on Android? `id` is a type on iOS. Will change it here, but primarily followed other method argument lists.

Comment (hansemannn) [inline android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java]: Updated.

File: android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java (hunk 3 above)
Comment (garymathews) [inline android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java]: And this :)

Comment (hansemannn) [inline android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java]: But `hasProperty` returns a boolean and `getProperty` the property itself. Please go ahead and edit it

Comment (garymathews) [inline android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java]: Yes, but you can only perform an `&&` operation on a `Boolean` in Java. So either:

```Java
if (child.hasProperty(TiC.PROPERTY_ID) && child.getProperty(TiC.PROPERTY_ID).equals(id)) {

// OR

if (child.getProperty(TiC.PROPERTY_ID) != null && child.getProperty(TiC.PROPERTY_ID).equals(id)) {
```

Comment (hansemannn) [inline android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java]: But that's how it was [before](https://github.com/appcelerator/titanium_mobile/pull/8677/commits/5df4ff7b1845481d17a5e21e2d1c5384ce8a2f8b#diff-bec22e2efdd8b352a7754ea42f435dc7L757) all that. You can edit it on Github, so we can ensure it doesn't cause problems again.

Comment (garymathews) [inline android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java]: I see, I think you might have missed my change to `hasProperty` [here](https://github.com/appcelerator/titanium_mobile/pull/8677#discussion_r94043594)

Comment (hansemannn) [inline android/titanium/src/java/org/appcelerator/titanium/proxy/TiViewProxy.java]: Gotcha, changed it. But the original implementation was like the second suggestion above already (except the string-casting), that's why I wondered. Sorry for the confusion, let's finish this 🙂

**Comment by garymathews** [comment]
CR: **PASS**
FT: **PASS**

👍

**Pull request merged**